core/src/main/resources/error/error-classes.json

```diff
@@ -75,6 +75,23 @@
       "The value <str> (<fmt>) cannot be converted to <targetType> because it is malformed. Correct the value as per the syntax, or change its format. Use <suggestion> to tolerate malformed input and return NULL instead."
     ]
   },
+  "DATATYPE_MISMATCH" : {
+    "message" : [
+      "Cannot resolve <sqlExpr> due to data type mismatch:"
+    ],
+    "subClass" : {
+      "BINARY_OP_DIFF_TYPES" : {
+        "message" : [
+          "the left and right arguments of the binary operator have different types (<left> and <right>)."
+        ]
+      },
+      "BINARY_OP_WRONG_TYPE" : {
+        "message" : [
+          "the binary operator requires the input type <inputType>, not <leftDataType>."
```

Comment:
Are you sure this is about the LEFT type?
Aren't those non-symetrical operators deriving the right type from the left type?

Reply:
First, the binary op checks that operators have the same type, and then check that the left is compatible with expected input type, see
https://github.com/apache/spark/blob/fab65f4e393359b9523e34457c877083dba0e7e5/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala#L765-L767

Reply:
I think this assumes the left and right are already the same type, maybe `<actualDataType>` is better here.

Reply:
I think you just introduce unnecessary assumptions between checks, and confuse users. The error message says **precisely** that LEFT is operand has wrong type. Users need to cast it or do something only with the LEFT one. When you says ACTUAL, does the users have to check and/or fix both operands, or left, or right?

Reply:
Where do we mention LEFT to the end users? The internal code calls `left.dataType` because the check above guarantees the left and right are the same type.

Reply:
Actually I wrote this line of code: https://github.com/apache/spark/pull/7420/files#r34704858

My intention was not to ask users to fix the left operand only.

Reply:
@cloud-fan Serge's comment has been addressed already.